### ISSUE TITLE
Cannot deserialize continuation with classes originating from a different classloader

### ISSUE THREAD

**Issue by jgomer2001** [mozilla/rhino#1178]
Hi, 

I'm making use of serializable continuations (version 1.7.14)

I have several Java classes that belong to a classloader other than the main classloader of my application. Sometimes, instances  that belong to such classes are part of a computation that I need to pause and resume later. 

Serialization of continuations work but when deserializing, I get a `ClassNotFoundException` because 
`NativeJavaObject#readObject` uses `Class#forName(String)` to set the `staticType` field. This call uses the classloader that specifically loaded `NativeJavaObject.class`...

Can you suggest a workaround for my case?

Also I wonder if setting `staticType` should involve `Class.forName` at all... I'm new to Rhino but maybe `javaObject.getClass()` would have the same effect...

Thanks in advance.

**Comment by p-bakker** [comment]
@szegedi @michbarsinai Happen to know the answer to this question?

**Comment by michbarsinai** [comment]
Not sure how you'd go about this inside Rhino (e.g. pre-populate the class cache to the like).
Using streams only, and assuming that's possible, you could use stubs. Upon serialization, create a stub that contains all the info needed to re-create the non-serializable object (that's the part I'm not sure about). On de-serialization, re-create these objects from the serialized stubs.

**Comment by szegedi** [comment]
It's always been one of the problems with serialization that upon deserialization, you need to be able to resolve class names to class objects – a typical job for a class loader. I can only recommend that you come up with a solution where class loader hierarchy used in deserialization is compatible to that of the one used in serialization, and you try to ensure that all your serialized runtime classes can be dereferenced by name in it. If you can't do that, you can _still_ decide to write a custom class loader that consults several independent class loaders when loading a class (effectively, having multiple parents.) This might or might not cause some additional unwanted consequences.

FWIW, back in 2005-2010 timeframe I wrote a continuation-based serializing distributed script executor that was driving hundreds of thousands of concurrently running long-executing workflow scripts at any given time, and the system even had a feature for on-demand loading and unloading of different coexisting versions of the code (so we could evolve our platform but still allow app developers to upgrade to newer versions of it when it suited them). This was all accomplished  with judicious use of phantom references and ordinary simple hierarchies of class loaders. Even for all that sophisticated usage, I still never had to resort to more exotic class loader architectures. I'm not saying "if it was enough for me it should be enough for you", but want to highlight that these can still get you very far, and that if you end up having some much more exotic class loader setups, you might want to look into whether you can reduce it back to something simpler.

**Comment by jgomer2001** [comment]
Really appreciate your feedback guys! I'm already following a strategy of stubs like Michael pointed out.  I'll dive deeper into this and share results here.

**Comment by p-bakker** [comment]
@michbarsinai @szegedi Tnx for your input!

> FWIW, back in 2005-2010 timeframe I wrote a continuation-based serializing distributed script executor that was driving hundreds of thousands of concurrently running long-executing workflow scripts at any given time, and the system even had a feature for on-demand loading and unloading of different coexisting versions of the code (so we could evolve our platform but still allow app developers to upgrade to newer versions of it when it suited them).

@szegedi is https://github.com/szegedi/spring-web-jsflow what you're referring to here?

As part of #954 I'd also like to add a page about Rhino's support for continuations. If any of you is willing to contribute some (basic) documentation, please let me know. Otherwise, I hope I can ask you to review whatever I've come up to when I get around to writting something down related to Continuations

**Comment by p-bakker** [comment]
@jgomer2001 am leaving this case open as a reminder to add something about this to our documentation, but besides that I don't think there's anything actionable in here, in the sense that I don't see a feature request or bug report in it, correct?

**Comment by jgomer2001** [comment]
Right, I didn't raise this as a direct request because I know serialization is a pain point and maybe it is more convenient for the project maintainers to re-define how serialization should work in the future as in #704 . Traditional Java serialization feels pretty uncomfortable.

> I'll dive deeper into this and share results here.

I was able to make something work, see: https://gist.github.com/jgomer2001/7acbe479bdffba6bcce8bfbb7122e314

not too beautiful but useful for me by now. This would actually allow me to use another serialization technology (eg. kryo) to some extent
